Cookie Store API docs updates - live examples (#38095)
mdn/content: files/en-us/web/api/cookiestore/index.md

## Examples

### Setting cookies

> The following example sets cookie by passing a `name` and `value` and then by setting an options value.

Reviewer: ```suggestion
The following example sets cookies by passing a `name` and `value` and then by setting an `options` value.
```

```html hidden
<button id="showCookies" type="button">Show cookies</button>
<button id="reset" type="button">Reset</button>
<pre id="log"></pre>
```

```css hidden
#log {
  height: 280px;
  overflow: scroll;
  padding: 0.5rem;
  border: 1px solid black;
}
```

```js hidden
const reload = document.querySelector("#reset");

reload.addEventListener("click", () => {
  window.location.reload(true);
});

const logElement = document.querySelector("#log");

function log(text) {
  logElement.innerText = `${logElement.innerText}${text}\n`;
  logElement.scrollTop = logElement.scrollHeight;
}
```

```js hidden
function logCookie(name, cookie) {
  if (cookie) {
    log(`${name}:`);
    for (const [key, value] of Object.entries(cookie)) {
      log(` ${key}: ${value}`);
    }
  } else {
    log(`${name}: Cookie not found`);
  }
}
```

#### JavaScript

> The example code sets two cookies, the first using `name` and `value` parameters, and the second using an "options" object with `name`, `value`, `expires`, and `partitioned` properties set.

Reviewer: ```suggestion
The example code sets two cookies, the first using `name` and `value` parameters, and the second using an `options` object with `name`, `value`, `expires`, and `partitioned` properties set.
```

Reviewer: But this is a weird way to put it. They both use an `options` object. I wouldn't say that the first uses "`name` and `value` parameters".

Author: Yes it is. Fixed in https://github.com/mdn/content/pull/38095/commits/3fadf3abceca2ce94f5c263d64b0753a1fc0008d

> The {{apiref("CookieStore.get()")}} method is then used to get the cookies, which are then logged.

Reviewer: ```suggestion
The {{domxref("CookieStore.get()")}} method is then used to get the cookies, which are then logged.
```

(but you could make it less passive with "We then use...")

Author: FYI, I modified your suggestion to be less passive.

Even if the values are not displayed, they are still set.

{{EmbedLiveSample('Setting cookies', 100, 350)}}

### Getting cookies

The following example shows how you can get a particular cookie using {{domxref("CookieStore.get()")}} or all cookies using {{domxref("CookieStore.getAll()")}}.

```html hidden
<button id="showCookies" type="button">Show cookies</button>
<button id="reset" type="button">Reset</button>
<pre id="log"></pre>
```

```css hidden
#log {
  height: 550px;
  overflow: scroll;
  padding: 0.5rem;
  border: 1px solid black;
}
```

```js hidden
const reload = document.querySelector("#reset");

reload.addEventListener("click", () => {
  window.location.reload(true);
});

const logElement = document.querySelector("#log");

function log(text) {
  logElement.innerText = `${logElement.innerText}${text}\n`;
  logElement.scrollTop = logElement.scrollHeight;
}
```

```js hidden
function logCookie(name, cookie) {
  if (cookie) {
    log(`${name}:`);
    for (const [key, value] of Object.entries(cookie)) {
      log(` ${key}: ${value}`);
    }
  } else {
    log(`${name}: Cookie not found`);
  }
}
```

#### JavaScript

> The example code first sets a number of cookies that we'll use for demonstrating the get methods.

Reviewer: ```suggestion
The example code first sets three cookies that we'll use for demonstrating the get methods.
```

First it creates `cookie1` and `cookie2` using the {{domxref("CookieStore.set()")}} method.

> Then it creates a couple of using the older synchronous {{domxref("Document.cookie")}} property (just so we can show that these are also fetched using our get methods)

Reviewer: ```suggestion
Then it creates a third cookie using the older synchronous {{domxref("Document.cookie")}} property (just so we can show that these are also fetched using the `get()` and `getAll()` methods).
```

    value: `cookie1-value`,
  });

  // Set a cookie passing an options object
  const day = 24 * 60 * 60 * 1000;
  await cookieStore.set({
    name: "cookie2",
    value: `cookie2-value`,
    expires: Date.now() + day,
    partitioned: true,
  });

  // Set cookie using document.cookie
  // (to demonstrate these are are fetched too)
  document.cookie = "favorite_food=tripe; SameSite=None; Secure";

  // Get named cookie and log properties
  const cookie1 = await cookieStore.get("cookie1");
  logCookie("cookie1", cookie1);

  // Get all cookies and log each
  const cookies = await cookieStore.getAll();
  if (cookies.length > 0) {
    log(`\ngetAll(): ${cookies.length}:`);
    cookies.forEach((cookie) => logCookie(cookie.value, cookie));
  } else {
    log("Cookies not found");
  }
}
```

Note that some logging and other code is omitted for brevity.

```js hidden
const showCookies = document.querySelector("#showCookies");

showCookies.addEventListener("click", () => {
  cookieTest();
});
```

#### Result

Press **Show cookies** to set the cookies and then display them in the log below.

{{EmbedLiveSample('Getting cookies', 100, 640)}}

### Delete a named cookie

> In this example, a cookie is deleted by passing its name to the `delete()` method.

Reviewer: ```suggestion
In this example, we delete a cookie by passing its name to the {{domxref("CookieStore.delete()","delete()")}} method.
```

```html hidden
<button id="showCookies" type="button">Show cookies</button>
<button id="reset" type="button">Reset</button>
<pre id="log"></pre>
```

```css hidden
#log {
  height: 100px;
  overflow: scroll;
  padding: 0.5rem;
  border: 1px solid black;
}
```

```js hidden
const reload = document.querySelector("#reset");

reload.addEventListener("click", () => {
  window.location.reload(true);
});

const logElement = document.querySelector("#log");

function log(text) {
  logElement.innerText = `${logElement.innerText}${text}\n`;
  logElement.scrollTop = logElement.scrollHeight;
}
```

```js hidden
// List cookies
async function getCookieNames() {
  let names = "";
  const cookies = await cookieStore.getAll();
  if (cookies.length > 0) {
    cookies.forEach((cookie) => (names += `${cookie.name} `));
  }
  return names;
}
```

#### JavaScript

The code first sets two cookies (which we then use to demonstrate deletion).

> We then list the names of both cookies (code for getting the cookie names not shown), delete one of the cookies, and then list all cooking names again.

Reviewer: ```suggestion
We then list the names of both cookies (code for getting the cookie names not shown), delete one of the cookies, and then list all cookie names again.
```

Author: I guess I was a bit hungry when I wrote this?

```js
async function cookieTest() {
  // Set two cookies
  await cookieStore.set({
    name: "cookie1",
    value: `cookie1-value`,
  });

  await cookieStore.set({
    name: "cookie2",
    value: `cookie2-value`,
  });

  // Log cookie names
  log(`Initial cookies: ${await getCookieNames()}`);

  // Delete cookie1
  await cookieStore.delete("cookie1");

  // Log cookie names again (to show cookie1 deleted)
  log(`Cookies after deleting cookie1: ${await getCookieNames()}`);
}
```

```js hidden
const showCookies = document.querySelector("#showCookies");

showCookies.addEventListener("click", () => {
  cookieTest();
});
```

Note that some logging and other code is omitted for brevity.

#### Result

Press **Show cookies** to run the code above.
This should show that the deleted cookie is present in the first list, and not in the second.

> {{EmbedLiveSample('Delete a named cookie', 100, 190)}}

Reviewer: Hm, interesting, for me the `favorite_food` cookie from the last example is still here in this one!

Author: Indeed. If you run the examples one after the other it will be present. However if you reset an example they are gone.

Do I need to fix this? I'm going to assume not - based on the fact that it still demonstrates deletion of the indicated cookie.

Reviewer: I think it is worth fixing, because it's confusing otherwise. As a reader I expect examples to be self-contained. I think it would be OK to have some delete-cookies code you run at the end, that you could even hide.

Author: @wbamberg Unfortunately I've run into problems, for which I need your advice.

1. First I tried to add this code - to do so I created some deletion code in https://github.com/mdn/content/pull/38095/commits/f3246bc2b4a6af4d7fa1f300d83c662791bc849d . 
   - When tested _locally_ this indicates that a cookie created using the `document.cookie` is not deleted. 
   - You can test this by doing the same test you that cause this discussion (`Getting cookies` before `Delete a named cookie` example 
   - the log code shows that cookie deletion succeeds for all the cookies - the method should throw if it can't delete.
   - BUT ... when I read the cookies it is still there.
   - Note that I left the logging code present just to demonstrate this.
   - Note, I use the word locally, because when I try on the PR such as [delete_a_named_cookie](https://pr38095.content.dev.mdn.mozit.cloud/en-US/docs/Web/API/CookieStore#delete_a_named_cookie) nothing is displayed - see next point.
2. When I try on the PR the examples are empty or indicate that the cookie does not exist.
   - I updated the `set()` example  [setting_a_cookie_with_options](https://pr38095.content.dev.mdn.mozit.cloud/en-US/docs/Web/API/CookieStore/set#setting_a_cookie_with_options)  to log the creation of the cookie.
   - On Chrome this gives and error `Error setting cookie2: UnknownError: Failed to execute 'set' on 'CookieStore': An unknown error occurred while writing the cookie.`
   - On FF NIghtly no error is raised - the set succeeds, but the cookie then is not read.
3. ON play for these examples the output is empty on Chrome and Gives me 404 text on  FF nightly.
4. Local test file used instead. This works if served from localhost:

   ```html
     <!DOCTYPE html>
      <html>
        <head>
          <meta charset="utf-8" />
          <title>Test CookieStore.set()</title>
          <style>
            #log {
              height: 180px;
              overflow: scroll;
              padding: 0.5rem;
              border: 1px solid black;
            }
      
            div:focus {
              border: 2px solid red;
            }
          </style>
        </head>
        <body>
          <h1>Test CookieStore.set()</h1>
      
          <button id="showCookies" type="button">Show cookies</button>
          <button id="reset" type="button">Reset</button>
          <pre id="log"></pre>
      
          <script>
            const reload = document.querySelector("#reset");
      
            reload.addEventListener("click", () => {
              window.location.reload(true);
            });
      
            const logElement = document.querySelector("#log");
      
            function log(text) {
              logElement.innerText = `${logElement.innerText}${text}\n`;
              logElement.scrollTop = logElement.scrollHeight;
            }
      
            function logCookie(name, cookie) {
              if (cookie) {
                log(`${name}:`);
                for (const [key, value] of Object.entries(cookie)) {
                  log(` ${key}: ${value}`);
                }
              } else {
                log(`${name}: Cookie not found`);
              }
            }
      
            async function cookieTest() {
              const day = 24 * 60 * 60 * 1000;
      
              try {
                // Set cookie
                await cookieStore.set({
                  name: "cookie2",
                  value: `cookie2-value`,
                  expires: Date.now() + day,
                  partitioned: true,
                });
              } catch (error) {
                log(`Error setting cookie2: ${error}`);
                console.log(error);
              }
      
              // Log the new cookie
              const cookie = await cookieStore.get("cookie2");
              logCookie("cookie2", cookie);
            }
      
            const showCookies = document.querySelector("#showCookies");
      
            showCookies.addEventListener("click", () => {
              cookieTest();
            });
          </script>
        </body>
      </html>
   ```
   - Note this file above fails with specific error if just loaded from file system, but only on Chrome (`Error setting cookie2: TypeError: Failed to execute 'set' on 'CookieStore': Cannot modify a secure cookie on insecure origin`) - on FF just doesn't work
   
I know that's a lot to digest. It seems to me that the unknown error is causing the problems. I suspect it is something to do with the PR server environment, but can't think what to try. Ideas?

Reviewer: I will look at this properly tomorrow, but just a note: none of these examples worked for me in the PR preview (and I saw the same errors you see), so I checked out your fork and ran it locally, and it all worked there. I don't know why (and probably should have mentioned it but forgot). Which I think is your point (2) right?

The page for [document.cookie](https://developer.mozilla.org/en-US/docs/Web/API/Document/cookie#notes) says:

> You can delete a cookie by updating its expiration time to zero.

Have you tried that?

Author: @wbamberg Regarding point 2 - yes "doesn't work in PRs and FF doesn't provide any debug about that".

I haven't tried deleting the cookie by updating its expiration time to zero. Kind of glad of that, because I think there are bugs here on deletion implementation. (I mean, yes I can do this, but we'll still be stuck.

Thanks for your help!

Reviewer: I don't have any answers, but I have more confirmation that something is weird. I wrote a little page to test: https://github.com/wbamberg/cookie-test. You can:

- add cookies, either using the cookie store or using document.cookie
- delete a cookie, delete all cookies, or list all cookies

If I run this locally using python's http.server, it works fine. If I run it from https://wbamberg.github.io/cookie-test/, I see exactly what you do: I can't delete cookies that are set using document.cookie.

I don't have any explanation for this yet.

Author: THanks very much @wbamberg - that's helpful. To me a clear indication that there are at least some bugs here. Enough to make it worth asking for an explanation in the FF implementation issue https://bugzilla.mozilla.org/show_bug.cgi?id=1937477#c14

Let's reconsider this later in the week.

Author: @wbamberg FYI the cookies created with  `document.cookies` are not being deleted because they don't have a matching path. This is because that method using the path of the path the cookie is created in, by default, while `set()` always uses `/` by default, and `delete()` matches against this.

This feels very painful because Firefox have chosen not to return values for cookies other than name and value in the `get()` and `getAll()` - so you can't reliably query for the paths to delete these. All you can do is either store the values or more likely always set the default if you have to use the document.cookies for whatever reasons. I guess it isn't really a problem if you're the one setting the code - you just won't use that API anymore.

The other problem is that there is no error if the cookie doesn't match. I'm checking that. 

Still working on why this doesn't work on the live site.

## Examples

The examples below can be tested by copying the code into a test harness and running it with a [local server](/en-US/docs/Learn_web_development/Howto/Tools_and_setup/set_up_a_local_testing_server), or deploying it to a website site such as GitHub pages.

<!-- The examples don't work as live examples in MDN environment (due to unknown errors) -->

### Setting cookies

This example shows how to set cookies by passing a `name` and `value`, and by setting an `options` value.

> The `cookieTest()` method first sets a cookie with `name` and `value` properties, while the second is set with `name`, `value`, and `expires` properties.

Reviewer: ```suggestion
The `cookieTest()` method sets one cookie with `name` and `value` properties and another with `name`, `value`, and `expires` properties.
```

We then use the {{domxref("CookieStore.get()")}} method to get each of the cookies, which are then logged.

```js
async function cookieTest() {
  // Set cookie: passing name and value
  try {
    await cookieStore.set("cookie1", "cookie1-value");
  } catch (error) {
    console.log(`Error setting cookie1: ${error}`);
  }

  // Set cookie: passing options
  const day = 24 * 60 * 60 * 1000;

  try {
    await cookieStore.set({
      name: "cookie2",
      value: "cookie2-value",
      expires: Date.now() + day,
      partitioned: true,
    });
  } catch (error) {
    log(`Error setting cookie2: ${error}`);
  }

  // Get named cookies and log their properties
  const cookie1 = await cookieStore.get("cookie1");
  console.log(cookie1);

  const cookie2 = await cookieStore.get("cookie2");
  console.log(cookie2);
}

cookieTest();
```

> Note that some browsers will only display the `name` and `value`, while others will display all the properties of the cookie.

Reviewer: Again I'm not sure we need to call this out here.

cookieTest();
```

The example should log "cookie1" and all three cookies separately.
One thing to note is that the cookie created using {{domxref("Document.cookie")}} may have a different path than those created using {{domxref("CookieStore.set()","set()")}} (which defaults to `/`).

### Delete a named cookie

This example shows how to delete a named cookie using the {{domxref("CookieStore.delete()","delete()")}} method.

The code first sets two cookies and logs them to the console.
We then delete one of the cookies, and then list all cookies again.

```js
async function cookieTest() {
  // Set two cookies
  try {
    await cookieStore.set("cookie1", "cookie1-value");
  } catch (error) {
    console.log(`Error setting cookie1: ${error}`);
  }

  try {
    await cookieStore.set("cookie2", "cookie2-value");
  } catch (error) {
    console.log(`Error setting cookie2: ${error}`);
  }

  // Log cookies
  console.log("Initial cookies");
  console.log(await cookieStore.getAll());

  // Delete cookie1
  await cookieStore.delete("cookie1");

  // Log cookies again (to show cookie1 deleted)
  console.log("Cookies after deleting cookie1");
  console.log(await cookieStore.getAll());
}
```

> When run in a console we should see the deleted cookie ("cookie1") is present in the first log array, and not in the second.

Reviewer: Why "when run in a console" specifically?

Author: Removed.
